operator-common/src/main/java/io/strimzi/operator/common/AbstractOperator.java

```diff
@@ -168,13 +182,61 @@ private String getLockName(String namespace, String name) {
+
+                if (cr.getSpec() == null)   {
+                    InvalidResourceException exception = new InvalidResourceException("Spec cannot be null");
+
+                    S status = createStatus();
+                    Condition errorCondition = new ConditionBuilder()
+                            .withLastTransitionTime(StatusUtils.iso8601Now())
+                            .withType("NotReady")
+                            .withStatus("True")
+                            .withReason(exception.getClass().getSimpleName())
+                            .withMessage(exception.getMessage())
+                            .build();
+                    status.addCondition(errorCondition);
+
+                    log.error("{}: {} spec cannot be null", reconciliation, cr.getMetadata().getName());
+                    updateStatus(reconciliation, status, unknownAndDeprecatedConditions).onComplete(statusResult -> {
+                        createOrUpdate.fail(exception);
+                    });
+
+                    return createOrUpdate.future();
+                }
+
                 log.info("{}: {} {} should be created or updated", reconciliation, kind, name);
-                return createOrUpdate(reconciliation, cr).recover(createResult -> {
-                    log.error("{}: createOrUpdate failed", reconciliation, createResult);
-                    return Future.failedFuture(createResult);
+
+                createOrUpdate(reconciliation, cr).onComplete(createOrUpdateResult -> {
+                    if (createOrUpdateResult.succeeded())    {
+                        updateStatus(reconciliation, createOrUpdateResult.result(), unknownAndDeprecatedConditions).onComplete(statusResult -> {
+                            if (statusResult.succeeded())   {
+                                createOrUpdate.complete();
+                            } else {
+                                createOrUpdate.fail(statusResult.cause());
+                            }
+                        });
+                    } else {
+                        if (createOrUpdateResult.cause() instanceof ReconciliationException) {
+                            ReconciliationException e = (ReconciliationException) createOrUpdateResult.cause();
+                            Status status = e.getStatus();
+
+                            log.error("{}: createOrUpdate failed", reconciliation, e.getCause());
+
+                            updateStatus(reconciliation, (S) status, unknownAndDeprecatedConditions).onComplete(statusResult -> {
+                                createOrUpdate.fail(e.getCause());
+                            });
+                        } else {
+                            log.error("{}: createOrUpdate failed", reconciliation, createOrUpdateResult.cause());
+                            createOrUpdate.fail(createOrUpdateResult.cause());
+                        }
+                    }
                 });
```

Comment:
This code doesn't actually work. When you use compose, you need to return `Future` and not just complete some `Promise` defined outside. So I don't think you can mix `onComplete` and `compose` so easily. I was thinking to rewrite it whole to `compose`, but it seemed lot more ugly than with `onComplete` mainly because these _special_ handlings where we return the original errors etc.